### ISSUE TITLE
Issue with model predictions

### ISSUE THREAD

**Issue by panakouris** [qubvel/segmentation_models#90]
I tried to train a PSPNet and a VPN for a multi-class segmentation task. When I try to predict on new data (I tried also to predict on the training data), the predictions seem to follow some of the patterns, but the problem is that the predicted image ends up to be of lower resolution, meaning that the same class is predicted for 4X4 or 8x8 blocks. For the task that I am working with, pixel prediction is quite important. Has anyone faced similar problem or have any advice? Thanks

**Comment by qubvel** [comment]
Hi @panakouris 
Could you describe your problem more detailed? May be with some code or picture examples? Your problem is not completely clear for me.

**Comment by panakouris** [comment]
Hi @qubvel,
I use remote sensing data where the input shape of the model is ```(384, 384, 6)``` and my labels after after converting them to one-hot-labels has ```(384, 384, 6)``` where 6 is the number of classes that I want to predict. 
I pre-processed the data by substracting the mean and dividing the std per channel.
I defined the model like:
``` model = PSPNet('resnet50', classes = num_classes, activation='softmax', input_shape=(384, 384, 6), encoder_weights=None, downsample_factor=4)```
Compiled it:
```model.compile(optimizer=adam, loss='categorical_crossentropy', metrics=['accuracy'])```
and fit it :
```model.fit(X_train, y_train, batch_size=32,
          epochs=200,
          validation_data=(X_val, y_val),
          callbacks=[logger, mcp, lr_decay]
)```
and finally saved it using ```model.save()```.
Then after loading the model and predicting on a patch I get the following:
Prediction :
![image](https://user-images.githubusercontent.com/24657463/56287501-c85c9500-611c-11e9-8481-359b16a464f1.png)
Truth:
![image](https://user-images.githubusercontent.com/24657463/56287545-ed510800-611c-11e9-860d-164967e8a100.png)
where you can see that the prediction is in a much lower resolution.

**Comment by panakouris** [comment]
I saw that this is probably something control by the downsample_factor. When I set it to 4, I got the following image of the same area:
![image](https://user-images.githubusercontent.com/24657463/56292992-89344100-6128-11e9-9031-ab0d6f392c96.png)

**Comment by qubvel** [comment]
Did you mean that image has lower `shape` or it is just look like it is lower resolution. PSPNet itself just collect `featues` from head of encoder, add PSP block (which does not upsample incoming featues to higher resolution) and than just apply interpolation for upsampling. 
So if you need pixel-wise accuracy try Unet architecture.

**Comment by panakouris** [comment]
Hi Pavel,
Yes I meant lower resolution. I will try with unet in order to achieve pixel-wise accuracy! Thanks for your answer!

**Comment by S4WRXTTCS** [comment]
Did you test out the Model right after training? So without saving and loading the model.

I'm asking because I'm having a similar issue with an FPN model. If I test out the model right after training it then it I get the expected results. 

I get pixelated results if I save the model, and then load the model in a different python program to run the inference.

It might be completely unrelated, but i thought it was worth mentioning just in case.

Here is running inference right after training
![result](https://user-images.githubusercontent.com/280576/57261114-61474780-701b-11e9-93a5-f9d68f71ce99.png)

Here is running inference after loading the model
![test_resultA](https://user-images.githubusercontent.com/280576/57261136-72905400-701b-11e9-84f8-bf958d1b0bfd.png)

Here is my code
[Code.zip](https://github.com/qubvel/segmentation_models/files/3150168/Code.zip)

**Comment by qubvel** [comment]
Hi, I think something wrong with resize layer, maybe interpolation changed after loading (nearest instead of bilinear).
Could you send me a model, I will check it out.

**Comment by S4WRXTTCS** [comment]
Here is the Model along with the input image, and the result I got without saving/loading the model.

https://drive.google.com/open?id=1uP-Tf5tlAwGmdILM9kh8UVKhsUFuJrGu

**Comment by qubvel** [comment]
Hi @S4WRXTTCS 
There was a bug, I hope I have fixed it.
Could you try to update the package and rerun your experiment? 
```bash
$ pip install -U git+https://github.com/qubvel/segmentation_models@fix-resize-image-layer-config
```

**Comment by S4WRXTTCS** [comment]
I'm getting the same pixelated results after updating the package, and re-running seg_test.py on my model.

At first I suspected that maybe there was an issue with the install so I tried it on a different machine that has never had this library, but I'm getting the same pixelization. 

Is there anything I'm missing in my seg_test.py code?

**Comment by qubvel** [comment]
You have to recreate model and load weights:
```model = FPN(...)
model.load_weights(path/to/previous/model.h5)
model.save(/path/to/updated/model.h5)```
or just train a new model.

**Comment by S4WRXTTCS** [comment]
I should have mentioned that I tried that as well.

I used the other computer to train a new model with this latest build, and I got the same results. Where the one that was run right after training looked good, and the one run on the saved model looks pixelated.

**Comment by S4WRXTTCS** [comment]
I tried recreating the model as you suggested within my seg_test.py program, and it works correctly. I should have thought about testing it that way since it literally recreates the model just like it was.  

But, I'm confused by why saving the entire model, and loading it still isn't working.  It's giving me the same pixelated output so it looks like I'm either saving it wrong, or loading it wrong. Or there is some bug somewhere.

**Comment by qubvel** [comment]
I will test it by myself, it is a bug with `ResizeImage` layer, I will find a way to fix it.

**Comment by qubvel** [comment]
For me updated package works.
It is a test you can run:

```python
import os
import keras
import numpy as np
import segmentation_models

os.environ['CUDA_VISIBLE_DEVICES'] = ''

model = segmentation_models.FPN(classes=1, activation='sigmoid')
# suppose you have train it here
# ...

# save model
model.save('./test_model.h5')

# load this model
reloaded_model = keras.models.load_model('./test_model.h5')

# make predictions with both models
pr_1 = model.predict(np.ones([1, 32, 32, 3]))
pr_2 = reloaded_model.predict(np.ones([1, 32, 32, 3]))

# compare results
np.allclose(pr_1, pr_2)
```
It is `failed` with previous version and `succeed` with new one. So make sure you update package from `fix-resize-image-layer-config` branch as I mentioned above.

**Comment by S4WRXTTCS** [comment]
Sorry about that.

The new version works correctly, and I get the good smooth results from the loaded model.

The reason it wasn't working earlier is I made a rookie mistake of having my seg_train.py file, and seg_test.py within the Segmentations Model git repository that I cloned.

I had it for reference, but completely forgot about the fact that it use that version instead of the installed version.  Using a different directory for seg_train.py, and seg_test.py allowed it to use the installed version of the segmentation_models library.
